### ISSUE TITLE
PL/SQL Type Object

### ISSUE THREAD

**Issue by MrUrdam** [NaturalDocs/NaturalDocs#67]
Hey,

I recently started using the (PL/)SQL Type Object. Using this kind of type, it is possible to create a class like construct that can later be used to create specific objects. 

Similar to a Package you have to define a Type definition and the body with the implementations. 

I would like to use natural docs in the definition, but because of the special way the type definition works I have problems to get the prototype parser to work. 

Here is a short example of a type object definition with two member functions/procedures and one constructor. 

In the type definition, the ending of the prototype is a comma (but not the comma in the parameters)

Everything starts with MEMBER. If I remove MEMBER the prototypes work but the SQL is invalid. 

Maybe adding MEMBER, MAP, ORDER, STATIC Anywhere in your prototype parter could resolve this. 

Constructor maybe is more special.

``` SQL
CREATE OR REPLACE TYPE O_LOGGER AS OBJECT (
  -- Variable: name
  -- The name of the logger used in the loggings. 
  name VARCHAR2(10 Byte),

  /*
   * Constructor: O_LOGGER
   * Creates a new <O_LOGGER> with the given name.
   *
   * Parameters:
   * self - The <O_LOGGER> object itself (Must not be included in call).
   * name - The name of the logger.
   * 
   * Returns:
   * The created object.
   */
  CONSTRUCTOR FUNCTION O_LOGGER(
    SELF IN OUT NOCOPY O_LOGGER,
    name               VARCHAR2
  ) RETURN SELF AS RESULT,

  /*
   * Procedure: staticTest
   * A example for a static procedure definition.
   */
  STATIC PROCEDURE staticTest,

  /*
   * Function: mapName
   * Can be used to compare o_loggers by its name.
   *
   * Returns:
   * A hash number of the name.
   */
  MAP MEMBER FUNCTION mapName RETURN NUMBER,

  /*
   * Function: orderName
   * Can be used to order o_loggers by its name.
   *
   * Returns:
   * A hash number of the name.
   */
  ORDER MEMBER FUNCTION ORDERName RETURN NUMBER,

  /*
   * Function: error
   * Logs the given error message with the given line number to the log files.
   *
   * Parameters:
   * line - The line where the error occurred.
   * message - The error message to log.
   *
   * Returns:
   * True if the error was logged.
   */
  MEMBER FUNCTION error(
    line     PLS_INTEGER,
    message  VARCHAR2
  ) RETURNS BOOLEAN,

  /*
   * Procedure: spacer
   * Logs a spacer line to the output file.
   */
  MEMBER PROCEDURE spacer
);
/
```

I hope you can kind of support the object definitions. 

Best regards

André.

**Comment by NaturalDocs** [comment]
Try this version:
https://www.naturaldocs.org/download/temp/Natural_Docs_temp6.zip

It better supports unexpected modifiers before functions and procedures, so it should handle things like MAP MEMBER now.  It also supports commas as prototype enders.  With this version I was able to run it on your example with no configuration changes and get the expected output.

The only change I had to make to the input was changing RETURNS to RETURN for error().  Microsoft's T-SQL uses RETURNS but I believe Oracle's PL/SQL only allows RETURN, correct?

**Comment by MrUrdam** [comment]
Thank you for your fast response and for the updated version.
The prototypes now work as expected in the type object. Looking great!

In Oracles PL/SQL as I know only the key RETURN is available in functions.
Is there any advanced functionality for the RETURN? Because it already was included in the prototype in recent versions.

I think someone already requested a feature, where the return type is also included in the documentation if a "Returns:" heading is included in the comment. Is there any plan to include/support this in the future?

Again thanks for helping me out so fast!

Best regards
Andre

**Comment by NaturalDocs** [comment]
I just wanted to know if RETURNS was a typo in your example or something I needed to support.  It sounds like it was a typo so everything is fine.

I'm not sure what you're referring to with requested feature.  Could you give me an example?

**Comment by MrUrdam** [comment]
It was just a typo. I didn't noticed it until now 🙈.

It would be great to have the returning datatype extracted from the prototype and listed just below the heading "Returns".

Similar to the parameter datatype but not below each parameter just below the heading. 


Instead of:

## Returns:
Returns the name...

Something like
## Returns:
###### Datatype
Returns the name...

**Comment by NaturalDocs** [comment]
That sounds like a good idea so I may try to implement it at some point, but it probably won't be soon.  The prototype parser doesn't extract the return value yet and would need to be updated for it.  It's definitely doable but it's not trivial.  There are other things I want to get to first and development is slow so it could be a while.

**Comment by NaturalDocs** [comment]
The fix in temp6.zip has been included in 2.2 Development Release 4.

**Comment by NaturalDocs** [comment]
[The fix has also been included in 2.2.](https://www.naturaldocs.org/download/)
